samza-core/src/main/scala/org/apache/samza/container/SamzaContainer.scala

```diff
@@ -462,6 +436,39 @@ object SamzaContainer extends Logging {
     else
       null
 
+    def maybeAsyncStreamTaskFactory : Object = {
+
+      val taskClassFactory = taskFactory match {
+        case null => TaskFactories.fromTaskClassConfig(config)
+        case _ => taskFactory
+      }
+
+      if (taskClassFactory == null) {
+        throw new SamzaException("Either the task class name or the task factory instance is required.")
+      }
+
+      val isAsyncTaskClass: Boolean = taskClassFactory.isInstanceOf[AsyncStreamTaskFactory]
+
+      if (isAsyncTaskClass) {
+        info("Got an AsyncStreamTask implementation.")
+      }
+
+      if (singleThreadMode && isAsyncTaskClass) {
+        throw new SamzaException("AsyncStreamTask cannot run on single thread mode.")
+      }
+
+      if (!singleThreadMode && !isAsyncTaskClass) {
+        info("Converting StreamTask to AsyncStreamTaskAdapter when running StreamTask w/ multiple threads")
+        new AsyncStreamTaskFactory {
```

Comment:
Forgot another factor: in standalone, when the taskFactory is an object that passed in, "fromConfig" actually does not make sense. Yet, the "maybeAsyncStreamTaskFactory" still applies to the passed in taskFactory as well. Hence, creating weird case that TaskFactories method would take an already instantiated taskFactory as the input parameter to generate the final task factory. Will try one more time to see whether that would clean up better.